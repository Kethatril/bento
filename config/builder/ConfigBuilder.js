const { Config } = require("webpack-config"),
    parse = require("./modules/parse");

module.exports = class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    constructor(options = {}) {
        this._options = parse(options);
        this._config = new Config();
        this._modules = [];
        this._manifest = {};
    }

    use(...modules) {
        this._modules = this._modules.concat(modules);
        return this;
    }

    build(fn) {
        this._modules.push(fn);

        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._modules,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            let config = new Config(),
                options = this._options;

            try {
                (function next(err) {
                    let async;

                    // Increment counter variable and skip any indices that
                    // don't exist. This allows sparse arrays to be iterated.
                    do {
                        ++i;
                    } while (!(i in modules) && i !== len);

                    if (err === false || err) return reject(err);

                    if (i === len) return resolve(config.toObject());

                    let result;

                    result = modules[i].call(
                        {
                            async() {
                                async = true;
                                return next;
                            }
                        },
                        config,
                        options
                    );
                    if (!async) {
                        next(result);
                    }
                })();
            } catch (e) {
                reject(e);
            }
        });
    }
};
