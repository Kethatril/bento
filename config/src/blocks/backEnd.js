import merge from "webpack-merge";

export default function(config, options) {
    if (options.get("proxy")) {
        config = merge(config, {
            devServer: {
                setup(app) {
                    app.use(
                        respMod({
                            rules: [
                                {
                                    match: new RegExp(
                                        options.get("proxy")["/"].target,
                                        "gi"
                                    ),
                                    replace: `//localhost:${options.get(
                                        "ports.webpack"
                                    )}`
                                }
                            ]
                        })
                    );
                }
            }
        });
    }

    if (options.get("html")) {
        const HtmlWebpackPlugin = require("html-webpack-plugin");

        config = merge(config, {
            plugins: [new HtmlWebpackPlugin(options.get("html"))]
        });
    }

    return config;
}
