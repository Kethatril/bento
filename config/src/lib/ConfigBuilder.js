import { parse, resolve, select } from "~/lib/utils";
import ManifestKeeper from "~/lib/ManifestKeeper";

export default class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    manifest = new ManifestKeeper();

    constructor(options = {}) {
        this._options = parse(resolve(options));
        this._blocks = options.blocks.list || [];
        this._timeout = options.blocks.timeout || 3000;
    }

    get options() {
        return this._options;
    }

    build() {
        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._blocks,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            const timeout = this._timeout;

            let config = {},
                options = this._options,
                utils = { select: select(options), manifest: this.manifest };

            (function next(result) {
                let async;

                // Increment counter variable and skip any indices that
                // don't exist. This allows sparse arrays to be iterated.
                do {
                    ++i;
                } while (!(i in modules) && i !== len);

                if (result === false || result instanceof Error)
                    return reject(err);

                if (i === len) {
                    return resolve(result);
                }

                try {
                    result = modules[i](result, options, utils);
                } catch (e) {
                    return reject(e);
                }
                if (!result || typeof result !== "object") {
                    throw new Error(
                        "[InvalidValueError] Each block must either return a new config object or a promise that resolves a new config object."
                    );
                }

                if (result.then) {
                    const id = setTimeout(function() {
                        if (!async) return;
                        return reject(
                            new Error(
                                `[TimeoutError] ${timeout} ms exceeded. For async blocks, ensure the returned promise is resolved.`
                            )
                        );
                    }, timeout);

                    const done = function(result) {
                        clearTimeout(id);
                        next(result);
                    };

                    async = true;
                    result.then(done, done);
                }

                if (!async) {
                    next(result);
                }
            })(config);
        });
    }
}
