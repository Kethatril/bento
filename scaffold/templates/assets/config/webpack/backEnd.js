module.exports = {
    <%_ if (kind === 'ssa') { _%>
    /**
     * ------------------------------------------------------------------------
     * Back-end proxy
     * ------------------------------------------------------------------------
     * For Server Side Application, map of endpoints that should be proxied to
     * a back-end server.
     * Note: currently, only "/" is supported.
     *
     */
    proxy: {
        "/": {
            target: <%- JSON.stringify(proxy) %>,
            changeOrigin: true,
            autoRewrite: true
        }
    }

    <%_ } else if (kind === 'spa') { _%>
    /**
     * ------------------------------------------------------------------------
     * Back-end HTML
     * ------------------------------------------------------------------------
     * Typically for Single Page Application, specify a template to generate
     * the index.html file
     *
     * This will pass straight to `html-webpack-plugin`
     *
     * @see https://github.com/jantimon/html-webpack-plugin#configuration
     *
     */
    html: {
        template: "./index.ejs",
        showErrors: true
    }
    <%_ } _%>
};
