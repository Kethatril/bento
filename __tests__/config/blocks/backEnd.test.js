import backEnd from "~/blocks/backEnd";
import ConfigBuilder from "~/lib/ConfigBuilder";

describe("blocks/backEnd", () => {
    const options = {
        ports: {
            browsersync: 3000,
            ui: 3001,
            webpack: 3002
        },

        proxy: {
            "/": {
                target: "http://localhost:8080",
                changeOrigin: true,
                autoRewrite: true
            }
        },
        blocks: {
            list: [backEnd]
        },
        html: {}
    };

    it("matches snapshot", () => {
        const builder = ConfigBuilder.create(Object.assign({}, {}, options));

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });
});
