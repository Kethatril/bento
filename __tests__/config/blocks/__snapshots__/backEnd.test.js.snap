// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks/backEnd matches snapshot 1`] = `
Object {
  "devServer": Object {
    "setup": [Function setup],
  },
  "plugins": Array [
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "minify": false,
        "showErrors": true,
        "template": "<cwd>/node_modules/html-webpack-plugin/default_index.ejs",
        "title": "Webpack App",
        "xhtml": false,
      },
    },
  ],
}
`;
